rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,162 @@
+
+Unlike PostgreSQL's `synchronous_commit` which offers multiple levels, RocksDB only provides a simple boolean option. This is because RocksDB is an embedded database and doesn't have the primary/standby architecture that PostgreSQL has.
+
+To optimize synchronous commit performance, RocksDB implements Group Commit, which is a common pattern in WAL-based systems. This mechanism batches multiple writes together into a single, larger WAL write and flush operation, should improve I/O throughput a lot when comparing with multiple small writes.
+
+(SlateDB can implement a similar Group Commit mechanism through its Commit Pipeline, allowing us to batch multiple writes into single WAL operations.)
+
+It worths to note that RocksDB defaults to `sync = false`, meaning WAL writes are not crash-safe by default.
+
+This default is likely to be a trade-off for performance. In many distributed systems (RocksDB's primary use case imo), some data loss on individual nodes is acceptable without compromising overall system durability. Examples include Raft clusters, distributed key-value stores, and stream processing state stores. For these use cases, enabling `manual_wal_flush` is often a good idea.
+
+RocksDB allows mixing writes with different sync settings. For example, if transaction A commits with `sync = false` and transaction B starts afterwards, transaction A's writes will be visible to readers in transaction B. When transaction B commits with `sync = true`, both transactions' writes are persisted. This ordering guarantee means that when a `sync = true` write commits, all previous writes are guaranteed to be persisted as well.
+
+Another important consideration is WAL write is possible to be failures. RocksDB handles these by retrying writes until determining whether the failure is temporary or permanent. In cases of permanent failure (like a full or corrupted disk), RocksDB marks the database state as fatal, rolls back the transaction like nothing happened, and switches the database instance to read-only mode.
+
+## Synchronous Commit in a summary 
+
+Based on the PostgreSQL and RocksDB references above, we can summarize the key semantics of Synchronous Commit:
+
+1. A write is only considered committed once the WAL has been persisted to storage. Until then, the data remains invisible to readers.
+2. If there is a permanent failure while persisting the WAL during a Synchronous Commit, the transaction rolls back completely. The database instance enters a fatal state and switches to read-only mode.
+3. Multiple levels of Synchronous Commit can be supported, allowing users to balance performance and durability requirements.
+4. Synchronous and Unsynchronous Commits can be interleaved in different transactions. A transaction using Synchronous Commit can read writes from transactions that used Unsynchronous Commit. When a Synchronous Commit persists, it also persists any previous Unsynchronous Commit writes in the WAL.
+
```

Comment:
One other thing I think we need to note here is, in the case of RocksDB, a reader doing a single `get()` can see a `sync=false` write that might disappear. There's no way for the reader to say they don't want to see those writes as far as I can tell. Similarly, in PostgreSQL, if `synchronous_commit=false`, readers will see writes that haven't been durably written, and again there's no way for readers to prevent that.